### PR TITLE
Configure CircleCI to update docker images on `master` commit 

### LINKED ISSUE #286: [CI] Have the CI build & update docker image on releases
Currently, the docker images on docker hub are so out-dated.
Automate the release pipeline (image build & upload)

### PR DESCRIPTION
**Pull request by red-robby**
This PR modifies the CircleCI configuration to automatically update Keystone's Docker images when a commit is made to the `master` branch. 

Due to dependencies between our tests and the Docker images, we divide the workflows into two groups: 

1. For the `master` branch, we update the Docker images and then run the tests _with the updated images_. 
2. For all other branches, we pull the Docker images and run the tests (no updates are made).

As CircleCI does not have a nice way of handling this at the workflow level, we create two versions of each job -- one that runs exclusively for the `master` branch and another that ignores `master` (i.e., runs for all other branches). 

Docker layer caching is enabled to reduce image build times (see [here](https://circleci.com/docs/docker-layer-caching/) for more details).

These Docker images are updated:
- `keystoneenclaveorg/keystone:master`
- `keystoneenclaveorg/keystone:init-rv32gc`
- `keystoneenclaveorg/keystone:init-rv64gc`

Resolves #286.

### CODE REVIEW

**Comment by dayeol** [comment]
how did you test it?

**Comment by red-robby** [comment]
@dayeol It was tested by writing to three test images (`citest0`, `citest1`, and `citest2`) in place of the actual images (`init-rv32gc`, `init-rv64c`, and `master`). These tests were successful. See [here](https://app.circleci.com/pipelines/github/keystone-enclave/keystone/248/workflows/87e0766f-828b-4a73-a899-7b471a26e0f3/jobs/1075), [here](https://app.circleci.com/pipelines/github/keystone-enclave/keystone/248/workflows/87e0766f-828b-4a73-a899-7b471a26e0f3/jobs/1077), and [here](https://app.circleci.com/pipelines/github/keystone-enclave/keystone/248/workflows/87e0766f-828b-4a73-a899-7b471a26e0f3/jobs/1078) for the test results. I confirmed they were actually pushed by checking Docker (though I've now deleted the test images).  

 The `master`/non-`master` functionality described above was tested by temporarily changing the functionality from `master` to `dev-circle-ci` (the branch I was developing on). When this change was made and the commit on `dev-circle-ci` took place, the test images were updated. See the test results mentioned above for confirmation. For non-master branches, the test on the most recent commit did _not_ effect the Docker image, as desired.

**Comment by grg-haas** [comment]
I really like this PR, and I would request that we actually *build* the docker image for each PR although we would only *push* it on commits to master. The reason for this is that currently PRs which make changes to the SDK are unable to run through the CI tests, since the SDK is prebuilt in the image (see for example my draft PR #317 ). Building the docker image each time would alleviate this issue, and then we could push the result only if it is a commit to master.

**Comment by dayeol** [comment]
@grg-haas
> at currently PRs which make changes to the SDK are unable to run through the CI tests, since the SDK is prebuilt in the image

Is this true? I thought it will be re-built and re-installed with this command:
https://github.com/keystone-enclave/keystone/blob/481822aad2a41ad85a7f17bf3e35115bbe129d9b/.circleci/config.yml#L30

Building docker image for each test is actually a good idea, but I would like to see how long does the full pipeline takes before call for it.

File: .circleci/config.yml [hunk 1]
```diff
@@ -262,3 +364,6 @@ workflows:
       - test:
           requires:
             - default-build
+  check-code-quality:
+    jobs:
+      - format
```
Comment (dayeol) [inline .circleci/config.yml:369]: please add a new line at the end of the file

**Review by dayeol — COMMENTED** [review]
Could you see if we can just build docker image for each PR, and skip pushing the images as @grg-haas has mentioned?
I if we could do that, we may not need make CI depend on the past images we built.

Thanks for the PR!

**Comment by grg-haas** [comment]
@dayeol I'm pretty sure this is an issue yes, see the circleCI build here (https://app.circleci.com/pipelines/github/keystone-enclave/keystone/182/workflows/d6594990-f550-43dd-acb7-85cb37d0aa9c/jobs/511/parallel-runs/0/steps/0-104) for example. Although my patch adds installations for more header files, this isn't triggered -- it almost seems as if the source isn't getting updated on clone in `/keystone`

**Review by dayeol — APPROVED** [review]

**Comment by grg-haas** [comment]
Agreed with all points, and sounds good to me

**Review by grg-haas — APPROVED** [review]

**Pull request merged**